README.md
# FavouriteLocations
 It is a application which helps to save your favourite locations.
 
 Backend:
 1.Login Authentication done using OAuth provided by Google's Firebase with Email and Password.
 2.The Location captured is send to the Database in Firebase .
 3.This application is completely hosted in Cloud Platform and uses Firebase as a simple backend.

 FrontEnd:
 1.ReactNative is used for the development.
 2.Package file contains:
 {
  "name": "FavouriteLocations",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "start": "node node_modules/react-native/local-cli/cli.js start"
  },
  "dependencies": {
  //These are the open source libraries used while development
    "firebase": "^3.3.0",
    "react": "15.2.1",
    "react-native": "0.31.0",
    "react-native-calendar-picker": "^3.0.0",
    "react-native-datepicker": "^1.3.2",
    "react-native-maps": "^0.7.1",
    "react-native-vector-icons": "^2.1.0",
    "reactfire": "^1.0.0"
  }
  }


ChangeLog
commit 963c61a3aed693fa97efde308404cd50f2853053
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Sun Jan 22 22:41:30 2017 +0530

    changed the markers animation color

commit 46b2faf2afe0b2e728ce1bc952656c00b1a0de67
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Thu Jan 12 13:25:22 2017 +0530

    Added animated markers

commit 71fdd5da9313f8bac0e3321ea9c428312a86bb93
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Thu Jan 12 11:03:20 2017 +0530

    Added annotations to the map

commit 0ae5df961e146b13c3d5d1808fce502b9806cd8c
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Wed Jan 11 18:21:04 2017 +0530

    UI changes for Iphone 5s and Iphone 6S Plus devices

commit f8fe4b06935bc315b9c52ea32d9c6376baae22dd
Author: ashoka.r <ashokicreate@gmail.com>
Date:   Sat Dec 31 18:49:16 2016 +0530

    can't find symblol issue resloved.

commit 4b39929046f4e89bb4ca7ac9c7db66d6d998645d
Author: ashoka.r <ashokicreate@gmail.com>
Date:   Sat Dec 31 18:40:31 2016 +0530

    Android: index.android.js: Loading Container Component as the First Component of FavouriteLocation app

commit 962748040cc35762cf57e723570cc136a04faa62
Author: ashoka.r <ashokicreate@gmail.com>
Date:   Sat Dec 31 18:29:21 2016 +0530

    react-native-router-flux added. Container Module added.

commit 499815e3e2cfc7f314b62ab54515d9311ba9e99e
Author: ashoka.r <ashokicreate@gmail.com>
Date:   Sat Dec 31 16:26:37 2016 +0530

    FavouriteLocations: code cleanup

commit 6781560e6450c9680bd54857c4b9b18dcf1d63ff
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Sat Dec 31 15:54:51 2016 +0530

    fixed firebase issue in favourite list screen

commit a53ccd72be861b86a7dece75b7320090409e3d0c
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Sat Dec 31 15:47:43 2016 +0530

    fixed FavListScreen issue

commit 9dfc1c25f82c0213a6a65bff673d6fa15e7a5684
Merge: 40cfc11 aa2d475
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Sat Dec 31 15:45:39 2016 +0530

    favList screen formatted

commit 40cfc11913270bbc087e91fce1b6922b884b8830
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Sat Dec 31 15:41:03 2016 +0530

    Formated Favourite Locations list screen

commit aa2d4750bd4b8d0ba42cb918d80ceea4726d35fc
Merge: ebfe4f9 802eccc
Author: ashoka.r <ashokicreate@gmail.com>
Date:   Sat Dec 31 15:32:09 2016 +0530

    Merge confilics resolved

commit ebfe4f902292dcf164cd72d8d1fd1f855901ffd4
Author: ashoka.r <ashokicreate@gmail.com>
Date:   Sat Dec 31 15:15:52 2016 +0530

    FavouriteLocationsFromHomeScreen: code clean up

commit 802eccc906b72122cf30f279d948ca414c1095da
Merge: 1c1be47 d2f4a20
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Sat Dec 31 14:35:44 2016 +0530

    fixed merge conflict

commit 1c1be47aa9cbb2ff0dea7adfc0f46809c0e16092
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Sat Dec 31 14:16:17 2016 +0530

    Formatted ForgotPasswordScreen and FavouriteLocationHomeScreen and moved images to folder

commit d2f4a2085138b9caed36f0330fa8b9cfa753f4b5
Author: ashoka.r <ashokicreate@gmail.com>
Date:   Sat Dec 31 13:31:35 2016 +0530

    ForgotPasswordScreen: code clean up

commit 5590f01f5390001d075698f6d9c329690b2354e7
Author: ashoka.r <ashokicreate@gmail.com>
Date:   Sat Dec 31 12:49:04 2016 +0530

    CreateAccountScreen: Code Cleanup

commit 68b42eb9740bcb1808fd8edd342a8c62d8eb9e3c
Author: ashoka.r <ashokicreate@gmail.com>
Date:   Sat Dec 31 09:49:28 2016 +0530

    HowToUseScreen: Code Clean Up

commit 0bb661238deec9b3e4493d13ab05fe1a58a67a0b
Author: ashoka.r <ashokicreate@gmail.com>
Date:   Sat Dec 31 07:15:43 2016 +0530

    Code Cleanup

commit 094c02e442ca41220ae44596226bf29bdafd1ede
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Sat Dec 31 00:14:53 2016 +0530

    Added new certificates and fixed Logout,back layout issues

commit f84c746d688f613011c77d43aba0167912912812
Merge: 4dc6626 fc1979f
Author: ashoka.r <ashokicreate@gmail.com>
Date:   Fri Dec 30 22:13:20 2016 +0530

    merge conflicts are resolved

commit 4dc66264514570cad43a135d3711c2de912c37ac
Author: ashoka.r <ashokicreate@gmail.com>
Date:   Fri Dec 30 21:47:47 2016 +0530

    react-native-loading-spinner-overlay component integrated.

commit fc1979f1315b8596c35cfe64be8ea8e556e358b1
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Fri Dec 30 10:55:56 2016 +0530

    Ad possition fixed

commit 1dc0c2028371eccdda28891c241c36a2fcbfabb5
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Fri Dec 30 10:39:51 2016 +0530

    Fixed ads possition in login screen

commit 85581979d2b50d1fa5792a82e0444a5366458407
Author: ashoka.r <ashokicreate@gmail.com>
Date:   Fri Dec 30 09:32:24 2016 +0530

    AdMobManager: Hard Coded string are removed.

commit e7a556094549b162f7d078e3d805bf914cae326c
Author: ashoka.r <ashokicreate@gmail.com>
Date:   Fri Dec 30 09:25:22 2016 +0530

    AdMobManger Module implemented.

commit 96ab8c0d3f18d294b77234328ff251476d2df537
Author: ashoka.r <ashokicreate@gmail.com>
Date:   Thu Dec 29 23:13:16 2016 +0530

    Google ads integration

commit 3e43dc28f56767ce6497c78bc5a5c33a4ca73c9d
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Thu Dec 29 19:12:39 2016 +0530

    Fixed share,map and delete buttons align issue

commit ec657fe2b41490b4d9e6daa4864a599219b9f79a
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Thu Dec 29 15:27:31 2016 +0530

    updtated the launch screen

commit 1b107a3ee536da3a4a7b6710b0a5475a5cfeac04
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Thu Dec 29 14:57:49 2016 +0530

    Fixed deleting locations issue

commit d74b960c74cce51c460662f4e0702ab6c29e01a6
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Tue Dec 27 15:41:58 2016 +0530

    Impleted removal of location from firebase

commit 7a259cbec17ef6e097c52da5fc28fdbc0941a7a9
Author: Vinay Hosamane K N <vhosb@allstate.com>
Date:   Tue Dec 27 00:53:05 2016 +0530

    Fixed shareOptions

commit b9cfe3239a0c69b238de4d8c218b12c63cdc3bab

components/Second.js
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 * @flow
 */

import React, { Component } from 'react';

import {
  StyleSheet,
  Text,
  View,
  TextInput,
  Navigator,
  NavigatorIOS,
  Image,
  Alert,
  StatusBar,
  TouchableOpacity
} from 'react-native';

import Button from 'react-native-button';
import MapView from 'react-native-maps'
import Geocoder from 'react-native-geocoder';

import Spinner from './Spinner.js';
import AdMobManager from './AdMobManager';
import HowToUseScreen from './HowToUseScreen'
import FavouriteLocationsFromHomeScreen from './FavouriteLocationsFromHomeScreen'

var bannerSize="smartBannerPortrait"
var testDeviceID="EMULATOR"
var adUnitID="ca-app-pub-3940256099942544/2934735716"

import * as firebase from 'firebase';
var config = {
    apiKey: "AIzaSyAFT0VK-KEMRiaqzlP2m1qpFodP1DNqm-8",
    authDomain: "my-favourite-locations.firebaseapp.com",
    databaseURL: "https://my-favourite-locations.firebaseio.com",
    storageBucket: "my-favourite-locations.appspot.com",
};
const firebaseApp = firebase.initializeApp(config);
const itemsRef = firebase.database();

  var newArray = [];
  var markers = [
 {
   latitude: 37.323,
   longitude:  -122.0527,
   title: 'Apple Head Quarters',
   subtitle: 'Cupertino'
 }
];

function randomColor() {
  return `#${Math.floor(Math.random() * 16777215).toString(16)}`;
}

export default class Second extends React.Component{
  constructor(props) {
      super(props)
      this.state = {
          name: 'MaxTech Login-->',
          MyAddress: 'Bangalore',
          usr_descrption: '',
          usr_placemark: '',
          position: {
              coords: {}
          },
          isLoading: false
      };
      console.disableYellowbox = true;
  }

  // this function will execute after rendering on the client occurs
  componentDidMount() {
    /*
        navigator is available via the Geolocation polyfill in React Native
        http://facebook.github.io/react-native/docs/geolocation.html#content

        navigator is the object through which you interact with the Geolocation interface

        *** Polyfill definition needs to be verified
        React Native allows for polyfills--code that provides functionality available in the browser, but
        that is not currently available in the runtime environment on mobile devices ***
        Geolocation is enabled by default when you create a project with react-native init.

        getCurrentPosition() and watchPostion() take a success callback, error callback, and options object
   */
   navigator.geolocation.getCurrentPosition(
       (position) => this.setState({position}),
       (error) => alert(error.message),
       {enableHighAccuracy: true, timeout: 20000, maximumAge: 1000}
    );
   navigator.geolocation.watchPosition((position) => {
       this.setState({position});
    });
  }

  _handleLogoutPress(event) {

    firebase.auth().signOut().then(function() {
        console.log("successfully logged out");
    }, function(error) {
        console.log(error);
    });
    this.props.navigator.pop();
  }

  _handleFavouriteLocationsAction(event) {
    this.setState({isLoading: true});
    this._navigate('Favourite_Locations_Clicked');
    // <CaptureLocationsScreen />
  }

  _handleCaptureLocationAction(event) {
    var username = this.state.username;
    var password = this.state.password;
    
    var description_Text = this.state.usr_descrption;
    var placemark_Text = this.state.usr_placemark;
    
    if(description_Text!='' || placemark_Text!='')
      {
         Alert.alert('Add My Location', 'Are you sure you want to add this to your Favourite locations list?.',
         [{text: 'No', onPress: () => console.log('Cancel Pressed!')},
          {text: 'Yes', onPress: () => {
            this.setState({isLoading: true});
            const userData = firebase.auth().currentUser;
            var userid = userData.uid;
            var myAddress = ''
            var NY = {lat: this.state.position.coords.latitude, lng: this.state.position.coords.longitude};
            Geocoder.fallbackToGoogle('AIzaSyCG887OvNRiQxuUoh3UYkKzxjcosHH1lRY');
             //{lat:12.985065,lng: 77.560499}
             // use the lib as usual
             let ret = Geocoder.geocodePosition(NY).then((res)=> {
                          myAddress = res["0"].formattedAddress
                          this.setState({MyAddress: myAddress},
                          function() {
                            console.log(this.state.MyAddress);
                            console.log('I am inside callback');
                          });
                          var dbRef = firebase.database().ref('testing/')
                          var savedbRef = dbRef.child(userid).push({
                                                Description : this.state.usr_descrption,
                                                Placemark : this.state.usr_placemark,
                                                Address: this.state.MyAddress,
                                                latitude: this.state.position.coords.latitude,
                                                longitude: this.state.position.coords.longitude
                                           })
                          console.log("dbRef "+dbRef)
                          console.log('savedbRef '+ savedbRef)
                          this.setState({isLoading: false});
                          
                       this.refs.description.setNativeProps({text: ''});
                       this.refs.placemark.setNativeProps({text: ''});
               
                       this.setState({usr_descrption:''});
                       this.setState({usr_placemark:''});
                          Alert.alert('Alert', 'Your Location saved successfully.', [{text: 'Thank you', onPress: () => console.log('OK Pressed!')}])
             })
           }
         }]
       )
      }
    
    else
      {
        this.setState({isLoading: false});
          Alert.alert('Alert', 'Please enter Description or Placemark for the location to improve accuracy.', [{text: 'Thank you', onPress: () => console.log('OK Pressed!')}])
      }
   
  }

  componentWillMount()
  {
    const userData = firebase.auth().currentUser;
    var userid = userData.uid;

    var itemsRef = firebase.database().ref('testing/'+userid);
    itemsRef.orderByChild(userid).on("child_added", (snapshot) => {
          console.log(snapshot.val());
          this.setState({isLoading: false});

          var data = snapshot.val()
          newArray.push(data)

          console.log(data.Description);
          console.log(data.latitude);
          console.log(data.longitude);
    }, (errorObject) => {
          console.log("The read failed: " + errorObject.code);
          this.setState({isLoading: false});
    });
  }

  _handleHowToUseAction(event) {
    this._navigate('How_To_Use_clicked');
  }

  
  onLocateClick()
  {
     navigator.geolocation.getCurrentPosition(
       (position) => this.setState({position}),
       (error) => alert(error.message),
       {enableHighAccuracy: true, timeout: 20000, maximumAge: 1000}
    );
   }
  
  _navigate(name, type='Normal') {
      if(name == 'How_To_Use_clicked') {
          this.props.navigator.push({
            component:HowToUseScreen,
            passProps: {name: name},
            type: type
          })
      } //end of if 'How_To_Use_clicked'
      if(name == 'Favourite_Locations_Clicked') {

        const userData = firebase.auth().currentUser;
        var userid = userData.uid;
        newArray = [];
         this.setState({isLoading: true});
        var itemsRef = firebase.database().ref('testing/'+userid);
        itemsRef.orderByChild(userid).on("child_added", (snapshot) => {
              console.log(snapshot.val());
              this.setState({isLoading: false});
 
              var data = snapshot.val()
              newArray.push(data)

              console.log(data.Description);
              console.log(data.latitude);
              console.log(data.longitude);
        }, (errorObject) => {
              console.log("The read failed: " + errorObject.code);
              this.setState({isLoading: false});
             // Alert.alert('Alert!', 'Please wait untill the connection with your cloud databsase is made. Try Again!', [{text: 'Thank you!', onPress: () => console.log('OK Pressed!')}])
        });

          if(newArray.length !=0) {
            this.setState({isLoading: false});
              this.props.navigator.push({
                  component:FavouriteLocationsFromHomeScreen ,
                  passProps: {name: newArray},
                  type: type
              })
          }
          else {
            this.setState({isLoading: false});
              Alert.alert('Alert!', 'Please wait untill the connection with your cloud databsase is made. Try Again!', [{text: 'Thank you!', onPress: () => console.log('OK Pressed!')}])
          }
        } //eof if 'Favourite_Locations_Clicked'
   }

   render() {
        return (
          <View style={styles.container_Second_2}>
            <Spinner visible={this.state.isLoading} size="large" color="white"/>
            <StatusBar backgroundColor="rgb(55,55,55)" barStyle="light-content"/>
            <View style={styles.halfHeight_Second_2}>
              <Button   style={styles.logoutButtonStyle}
                      onPress={this._handleLogoutPress.bind(this)}>
                      Logout 🔐
               </Button>
            </View>

             <View style={styles.quarterHeight1_Second_2}>
                <MapView style={styles.quarterHeight1_Second_2}
                         showsUserLocation={true}
                         followUserLocation={true}
                         mapType='standard'
                         region={{latitude: this.state.position.coords.latitude, latitudeDelta: 0.009, longitude: this.state.position.coords.longitude, longitudeDelta: 0.009}}
                         zoomEnabled={true}
                         scrollEnabled={true}
                         showsScale={true}
                         >  

                </MapView>
                <TouchableOpacity onPress={()=>{
                           //Share.open(shareOptions);
                           this.onLocateClick();
                           }}>
                  <View style={{width: 30, height: 30}}>
                     <Image source={require('../images/LocateIcon.png')}  style={styles.backgroundImageShare}></Image>
                 </View>
                 </TouchableOpacity>
               
             </View>

             <View style={styles.quarterHeight2_Second_2}>
               <View style={{flex: 0.80, justifyContent: 'space-around', alignItems: 'center'}}>
                <TextInput ref="description"
                           style={{height: 35, borderColor: 'gray', borderWidth: 1 , marginTop: 6 , padding : 10 , marginLeft : 5 , marginRight : 5}}
                           placeholder= "Please describe location EX.My Home,Best Gobi Manchuri"
                           placeholderTextColor = 'black'
                           returnKeyType = {"next"}
                           autoFocus = {false}
                           autoCapitalize = "none"
                           autoCorrect = {false}
                           clearButtonMode = 'while-editing'
                           onChangeText={(text) => {
                             this.setState({usr_descrption:text});
                           }}
                           onSubmitEditing={(event) => {
                             this.refs.placemark.focus();
                           }}
                />
                <TextInput ref="placemark"
                          style={{height: 35, borderColor: 'gray', borderWidth: 1 , marginTop: 6 , padding : 10 , marginLeft : 5 , marginRight : 5}}
                          placeholder= "Please give Placemark Ex.Near Vijaya Bank,Behind Bus Stand"
                          placeholderTextColor = 'black'
                          returnKeyType = {"done"}
                          autoFocus = {false}
                          autoCapitalize = "none"
                          autoCorrect = {false}
                          clearButtonMode = 'while-editing'
                          onChangeText={(text) => {
                            this.setState({usr_placemark:text});
                          }}
                          onSubmitEditing={(event) => {
                          }}
                />
                <Button
                style={{fontSize: 20}}
                onPress={this._handleCaptureLocationAction.bind(this)}>
                        Capture Location
                </Button>
                <Button
                style={{fontSize: 20}}
                onPress={this._handleFavouriteLocationsAction.bind(this)}>
                          Favourite Locations
                </Button>
                <Button
                style={{fontSize: 20}}
                onPress={this._handleHowToUseAction.bind(this)}>
                          About App
                </Button>
              </View>
                <View style={styles.adStyle}>
                  <AdMobManager  bannerSize = {bannerSize} testDeviceID = {testDeviceID} adUnitID = {adUnitID}/>
                </View>
            </View>

          </View>
   );
 }
}

const styles = StyleSheet.create({
    container_Second_2: {
        flex: 1,
        flexDirection: 'column'
    },
    halfHeight_Second_2: {
        flex: .09,
        backgroundColor: '#FF3366',
        justifyContent: 'center'
    },
    logoutButtonStyle: {
        borderColor: 'white',
        color: 'white',
        alignSelf: 'flex-end'
    },
    quarterHeight1_Second_2: {
        flex: .46,
        //backgroundColor: '#fffaf0'
      backgroundColor: '#CCC',
    },
    quarterHeight2_Second_2: {
        flex: .45,
        backgroundColor: '#CCC',
    },
    adStyle: {
        flex: 0.20,
        justifyContent: 'flex-end',
        alignItems: 'flex-end',
    },
   backgroundImageShare: {
        marginTop: 1,
        width: 40,
        height: 30,
        marginRight: 50,
        alignItems: 'flex-end'
    },
});
